### PR TITLE
Perf: Better response times from api/public/v1/courses

### PR DESCRIPTION
**Pull request by inulty-dfe**
## Context

Counting the number of courses in a recruitment cycle is very expensive. We do this on every request to the courses endpoint.

### This is an "agonising" endpoint in Skylight

image

## Loadtesting the change
The tests were run with Siege.

[Siege documentation](https://linux.die.net/man/1/siege) 

All tests were run against review apps. The review apps all had 1 replica with 512MiB memory.

### Before changes

```
inulty-dfe ~/Code/publish-teacher-training ( perf-api-providers) $ siege -c 10 -t 1m -if app/controllers/api/loadtest-urls-other-branch.txt

{       "transactions":                           91,
        "availability":                       100.00,
        "elapsed_time":                        59.50,
        "data_transferred":                    19.95,
        "response_time":                        5.97,
        "transaction_rate":                     1.53,
        "throughput":                           0.34,
        "concurrency":                          9.13,
        "successful_transactions":                31,
        "failed_transactions":                     0,
        "longest_transaction":                 19.52,
        "shortest_transaction":                 0.97
}71
}
```

### With just `loaded?` changes

```
inulty-dfe ~/Code/publish-teacher-training ( perf-api-providers) $ siege -c 10 -t 1m -if app/controllers/api/loadtest-urls-no-cache.txt

{       "transactions":                           75,
        "availability":                       100.00,
        "elapsed_time":                        59.49,
        "data_transferred":                    26.39,
        "response_time":                        7.16,
        "transaction_rate":                     1.26,
        "throughput":                           0.44,
        "concurrency":                          9.03,
        "successful_transactions":                45,
        "failed_transactions":                     0,
        "longest_transaction":                 19.76,
        "shortest_transaction":                 1.11
}
```

### With count query cached and `loaded?` changes

```
inulty-dfe ~/Code/publish-teacher-training ( perf-api-providers) $ siege -c 10 -t 1m -if app/controllers/api/loadtest-urls.txt

{       "transactions":                           94,
        "availability":                       100.00,
        "elapsed_time":                        59.86,
        "data_transferred":                    51.11,
        "response_time":                        6.07,
        "transaction_rate":                     1.57,
        "throughput":                           0.85,
        "concurrency":                          9.53,
        "successful_transactions":                84,
        "failed_transactions":                     0,
        "longest_transaction":                 13.18,
        "shortest_transaction":                 1.90
}
```

### Result

|Operation|Original|No cache changes|Cache changes|
|---|---|---|---|
| throughput |  0.34 | 0.44 | 0.85
| successful_transactions| 31 | 45 | 84 |
| data_transferred | 19.95 | 26.39 | 51.11|

### The URLs
The URLs used were variations on this list.
I added `no_cache` for the "no cache" test run.
I changed the PR number for the "original" test run.

image

## Changes proposed in this pull request

- Use `loaded?` to avoid making requests to the database when we already have records in memory.
- Cache the course record count per recruitment cycle for a period of time.
- The `filter` field is not used in the cache key because it is ignored in the query. The API does not filter courses.

## Guidance to review

- How long should the course count be cached? I would suggest 1 day.

## Things to check

- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] Required environment variables have been updated added to the Azure KeyVault
- [ ] Inform data insights team due to database changes
- [ ] Make sure all information from the Trello card is in here
- [ ] Rebased main
- [ ] Cleaned commit history
- [ ] Tested by running locally
- [ ] Attach PR to Trello card

### CODE REVIEW

File: app/controllers/api/public/v1/courses_controller.rb [hunk 1]
```diff
@@ -7,24 +7,40 @@ class CoursesController < API::Public::V1::ApplicationController
...
 
         private
 
+        def cached_course_count
+          year = permitted[:recruitment_cycle_year] || RecruitmentCycle.current.year
+
+          if permitted[:no_cache]
+            courses.count('course.id')
+          else
+            Rails.cache.fetch("api_course_count_#{year}", expires_in: 5.minutes) do
```
Comment (Nitemaeric) [inline app/controllers/api/public/v1/courses_controller.rb:24]: We can let Rails handle the stringification our cache key here for our convenience. Not a blocker.

```suggestion
            Rails.cache.fetch([:api_course_count, year], expires_in: 5.minutes) do
```

File: app/controllers/api/public/v1/courses_controller.rb [hunk 2]
```diff
@@ -7,24 +7,40 @@ class CoursesController < API::Public::V1::ApplicationController
...
           render json: { status: 400, message: 'Invalid changed_since value, the format should be an ISO8601 UTC timestamp, for example: `2019-01-01T12:01:00Z`' }.to_json, status: :bad_request
         end
 
         private
 
+        def cached_course_count
+          year = permitted[:recruitment_cycle_year] || RecruitmentCycle.current.year
+
+          if permitted[:no_cache]
+            courses.count('course.id')
```
[the remark this answers is not in the document]
Comment (inulty-dfe) [reply app/controllers/api/public/v1/courses_controller.rb:22]: Yeah good point, I saw a discussion in `DfE::Analytics` around this too. I would be interested to see if we could optimise the query but I puropsely avoided getting into that here.

File: app/models/course.rb [hunk 3]
```diff
@@ -585,6 +585,8 @@ def gcse_grade_required
   end
 
   def last_published_at
+    return enrichments.select(&:last_published_timestamp_utc).max_by(&:last_published_timestamp_utc)&.last_published_timestamp_utc if enrichments.loaded?
+
     enrichments.maximum(:last_published_timestamp_utc)
   end
```
Comment (Nitemaeric) [inline app/models/course.rb:591]: This looks like it can be generalised for all associations.

Have you considered looking into [ActiveRecord/Associations/CollectionProxy](https://api.rubyonrails.org/v7.2/classes/ActiveRecord/Associations/CollectionProxy.html)? A `#maximum` method there might be able to address this issue across the entire app with a single change.

File: app/controllers/api/public/v1/courses_controller.rb [hunk 4]
```diff
@@ -7,24 +7,40 @@ class CoursesController < API::Public::V1::ApplicationController
...
         rescue ActiveRecord::StatementInvalid
           render json: { status: 400, message: 'Invalid changed_since value, the format should be an ISO8601 UTC timestamp, for example: `2019-01-01T12:01:00Z`' }.to_json, status: :bad_request
         end
 
         private
 
+        def cached_course_count
+          year = permitted_params[:recruitment_cycle_year] || RecruitmentCycle.current.year
+
+          if permitted_params[:no_cache]
```
Comment (Nitemaeric) [inline app/controllers/api/public/v1/courses_controller.rb:21]: Below might be something for another PR.

It might be more conventional to read the `Cache-Control` request header and return cached vs non-cached results.

I haven't quite dug into if that's indeed the right header, or what the header value should be (I _think_ `no-cache`?), and also if my assumption is considered standard practice, but it might be worth looking into.

Comment (inulty-dfe) [reply app/controllers/api/public/v1/courses_controller.rb:21]: I'm happy to remove this, I don't think it should be merged. Undocumented api params probably aren't a good idea.

As for the cache headers, I believe those (response headers) are for instructing how the client should cache, not how the server should respond.

File: app/models/course.rb (hunk 3 above)
Comment (inulty-dfe) [reply app/models/course.rb:591]: Yeah that looks tempting. but I would be careful about doing that. It's something we should discuss though, it seems like a useful thing to do.

File: app/controllers/api/public/v1/courses_controller.rb (hunk 4 above)
Comment (Nitemaeric) [reply app/controllers/api/public/v1/courses_controller.rb:21]: I _think_ that [this section](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#cache_directives) is indicating that there are request header standards too, but yeah, I haven't seen it in practice yet. I wonder how the browser implements the "Disable cache" checkbox.

But yeah, I agree. Let's remove the `no_cache` for now.

File: app/controllers/api/public/v1/courses_controller.rb [hunk 5]
```diff
@@ -7,24 +7,36 @@ class CoursesController < API::Public::V1::ApplicationController
...
         rescue ActiveRecord::StatementInvalid
           render json: { status: 400, message: 'Invalid changed_since value, the format should be an ISO8601 UTC timestamp, for example: `2019-01-01T12:01:00Z`' }.to_json, status: :bad_request
         end
 
         private
 
+        def cached_course_count
+          year = permitted_params[:recruitment_cycle_year] || RecruitmentCycle.current.year
+
+          Rails.cache.fetch("api_course_count_#{year}", expires_in: 5.minutes) do
```
Comment (tomas-stefano) [inline app/controllers/api/public/v1/courses_controller.rb:21]: Let's have a discussion about this expiration time?

**Review by tomas-stefano — APPROVED** [review]
Everything looks good. Great initiative. ✌️ 

Let's have a separate discussion about the cache & expiration.
